### ISSUE TITLE
Continuously rebase upstream Dex so we can stay up-to-date

### ISSUE THREAD

**Issue by vito** [concourse/concourse#6007]
...and notice merge conflicts.

Users are waiting on things which have been fixed/implemented upstream, and there may also be important security fixes that we miss as long as our fork strays behind.

Ideal outcome:

1. Whenever a new version of Dex is shipped from `dexidp/dex`, all of our PR branches are rebased on it and force-pushed.
2. A new branch is created from upstream `master`, and all of our newly rebased PR branches are merged into it.
3. A commit is made on the new branch that rewrites all imports from `github.com/dexidp/dex` to `github.com/concourse/dex`.
3. This new branch containing upstream + all of our rebased PRs is tagged and pushed.
4. This new branch is `git push -f`ed over `master`.
5. The `concourse/dex` dependency in `concourse/concourse` is bumped to the new tag pushed from step 3.

Various scripts for this have already been created on the `maintenance` branch, but it looks like knowledge of this branch was lost and it hasn't been used in a while.

* Step 1 can be automated with the `./update-all-prs` and `./push-all-prs` scripts.
  * Note that `./update-all-prs` takes care to keep track of inter-PR dependencies, i.e. that PR 2 needs to be based on PR 1.
* Steps 2 and 3 can be automated with the `./update-merged-branch` script.
  * This script merges all PRs into a freshly created `master` branch, reset to upstream `master`.
  * It then tags HEAD as `merged-$(date "+%Y%m%d%H%M%S")` so that there is a permanent reference that `go.mod` can point to that will survive `git push -f`s.
* Step 4 can be done with a `put` step.
  * **Note that there appear to be changes made directly to `master` which may need to be put into a PR so that they don't get lost.**
* Step 5 has never been automated - now's a good time!

**Comment by xtremerui** [comment]
To make step 1 easier we have cleaned up all branches with `pr/*` and removed those that were merged into upstream already. So the script for step 1 will just try to update all existing synced but not merged PRs.

One exception is pr/http-filesystem. There is an existing PR that based on this branch https://github.com/dexidp/dex/pull/1416.  I can't sync the branch and rename it with tailing `sync` in case it closes the existing upstream PR. We will keep this branch and be noted that it is already synced with upstream.

**Comment by xtremerui** [comment]
we need to make these commits that merged in master to its own PR(or put into an existing pr/* branch) so after the new master tag is created we dont lose any of those changes.

- sec gitlab name fix https://github.com/concourse/dex/commit/821b48abfd886ddbabf609e9e5b1dfa5fb1490cd
- cf connector pagination https://github.com/concourse/dex/pull/6

**Comment by xtremerui** [comment]
Related work is done in https://github.com/concourse/dex/tree/maintenance.
